### PR TITLE
Fix generic attribute support

### PR DESCRIPTION
**Pull request by Pessimistress**
- Fix: `VertexArray.setGeneric`
- New: `isGeneric` flag in attribute definition. If `true`, treat attribute as generic. `attribute.value` must be a typed array of size 1-4 in this case.

### CODE REVIEW

File: src/webgl/program.js [hunk 1]
```diff
@@ -162,14 +163,21 @@ export default class Program extends Resource {
       const buffer = buffers[bufferName];
       // DISABLE MISSING ATTRIBUTE
       if (!buffer) {
-        this.vertexAttributes.disable(location);
-      } else {
+        // do not disable attribute at location 0
+        if (location > 0) {
```
Comment (Pessimistress) [inline src/webgl/program.js]: Moved from `disable`

File: src/webgl/vertex-array.js [hunk 2]
```diff
@@ -165,12 +166,9 @@ export default class VertexArray extends Resource {
 
   // Disable an attribute
   disable(location) {
-    // Don't disable location 0
```
Comment (ibgreen) [inline src/webgl/vertex-array.js]: Nit: I see... Seems to me like this obscure condition still somehow does belong in disable(). Maybe add a flag to disable(.., {disableZero = false} = {}) instead?

File: src/webgl/vertex-array.js [hunk 3]
```diff
@@ -221,20 +219,48 @@ export default class VertexArray extends Resource {
 
     switch (array.constructor) {
     case Float32Array:
-      gl.vertexAttrib4fv(location, array);
+      this._setGenericFloatArray(location, array);
```
Comment (ibgreen) [inline src/webgl/vertex-array.js]: Should we drop the warning at the top of this function.

File: src/webgl/vertex-array.js [hunk 4]
```diff
@@ -221,20 +219,48 @@ export default class VertexArray extends Resource {
...
       break;
     case Int32Array:
       assert(isWebGL2(gl));
-      gl.vertexAttribI4iv(location, array);
+      this._setGenericIntArray(location, array);
```
Comment (ibgreen) [inline src/webgl/vertex-array.js]: Nit; Move the asserts into the helpers?

File: src/webgl/vertex-array.js [hunk 5]
```diff
@@ -36,6 +36,7 @@ const PARAMETERS = [
   GL_VERTEX_ATTRIB_ARRAY_DIVISOR
 ];
 
+const ERR_GENERIC_ATTRBUTE_LENGTH = 'vertex attribute size must be between 1 and 4';
```
Comment (ibgreen) [inline src/webgl/vertex-array.js]: Typo ATTRIBUTE.

File: src/webgl/vertex-array.js [hunk 6]
```diff
@@ -221,20 +219,48 @@ export default class VertexArray extends Resource {
...
+  }
 
-    return this;
+  _setGenericFloatArray(location, array) {
+    switch (array.length) {
+    case 1: this.gl.vertexAttrib1fv(location, array); break;
+    case 2: this.gl.vertexAttrib2fv(location, array); break;
+    case 3: this.gl.vertexAttrib3fv(location, array); break;
+    case 4: this.gl.vertexAttrib4fv(location, array); break;
+    default: throw new Error(ERR_GENERIC_ATTRBUTE_LENGTH);
```
Comment (ibgreen) [inline src/webgl/vertex-array.js]: Nit: Could just `assert(false)` per previous discussions.

File: test/webgl/vertex-array.spec.js [hunk 7]
```diff
@@ -57,9 +57,7 @@ test('WebGL#VertexAttributes#enable', t => {
     vertexAttributes.disable(i);
   }
 
-  t.equal(vertexAttributes.getParameter(GL.VERTEX_ATTRIB_ARRAY_ENABLED, {location: 0}), true,
-    'vertex attribute 0 should **NOT** be disabled');
-  for (let i = 1; i < MAX_ATTRIBUTES; i++) {
+  for (let i = 0; i < MAX_ATTRIBUTES; i++) {
```
Comment (ibgreen) [inline test/webgl/vertex-array.spec.js]: See comments above. The received wisdom is that WebGL breaks on some older platforms if we disable attribute 0 so we should avoid it when we can.

Comment (Pessimistress) [reply test/webgl/vertex-array.spec.js]: Generic attribute will not work if the attribute is not disabled. Does this mean that we need to warn user if they attempt to `setGeneric` on location 0?

Comment (ibgreen) [reply test/webgl/vertex-array.spec.js]: > Does this mean that we need to warn user if they attempt to setGeneric on location 0?

I am not sure. It is possible that the platforms that have an issue with disabling location 0 do not have that issue if a generic attribute has been set on that location. 

Most apps don't use generic attributes so I wouldn't worry about issuing a warning.

**Pull request merged**